src/esm.ts

```diff
@@ -147,12 +149,40 @@ export function createEsmHooks(tsNodeService: Service) {
       return defer();
     }
 
-    // pathname is the path to be resolved
+    let candidateSpecifiers: string[] = [specifier];
+
+    if (context.parentURL) {
+      const parentUrl = new URL(context.parentURL);
+      if (parentUrl.pathname && extname(parentUrl.pathname) === '.ts') {
+        const mappedSpecifiers = mapPath(specifier);
+        if (mappedSpecifiers) {
+          candidateSpecifiers = mappedSpecifiers.map((path) =>
+            pathToFileURL(path).toString()
+          );
+        }
+      }
+    }
+
+    let candidateSpecifier: string | undefined;
+    while ((candidateSpecifier = candidateSpecifiers.shift())) {
```

Comment:
Personal nit: prefer `for()` loop to mutating array.

Reply:
https://github.com/TypeStrong/ts-node/pull/1585/commits/2fcb9821593fd9235f45585ed0c95b1943a9c3de